File: solutions/clojure/phone-number/1/src/phone_number.clj
(ns phone-number)

(defn number [num-string]
  (let [digits (apply str (filter #(Character/isDigit %) num-string))
        first-digit (first digits)
        first-exchange-digit (nth digits 3)
        len (count digits)]
    (cond (and (= 11 len)(= \1 first-digit)) (subs digits 1)
          (or
           (= \0 first-digit)
           (= \1 first-digit)
           (= \0 first-exchange-digit)
           (= \1 first-exchange-digit)
           (= 11 len)
           (= 9 len)) "0000000000"
          :else digits)))

(defn area-code [num-string]
  (apply str (take 3 (number num-string))))

(defn pretty-print [num-string]
  (let [number (number num-string)
        area (subs number 0 3)
        triplet (subs number 3 6)
        rest (subs number 6)]
    (str "(" area ") " triplet "-" rest)))

File: solutions/go/phone-number/1/phone_number.go
package phonenumber

import (
    "errors"
    
)

func Number(phoneNumber string) (string, error) {
	digits := []rune{}
    for _, ch := range phoneNumber {
        if ch >= '0' && ch <= '9' {
            digits = append(digits, ch)
        }
    }
    if len(digits) == 11 {
    	if digits[0] != '1' {
            return "", errors.New("Invalid number")
        }
        digits = digits[1:]
    }
    if len(digits) != 10 {
        return "", errors.New("Invalid number of digits")
    }
    if digits[0] == '0' || digits[0] == '1' {
        return "", errors.New("Invalid code")
    }
    if digits[3] == '0' || digits[3] == '1' {
        return "", errors.New("Invalid exchange")
    }
    return string(digits), nil
}

func AreaCode(phoneNumber string) (string, error) {
	num, err := Number(phoneNumber)
    if err != nil {
        return "", err
    }
    return num[0:3], err
}

func Format(phoneNumber string) (string, error) {
	num, err := Number(phoneNumber)
    if err != nil {
        return "", err
    }
    fmt := []rune{}
    for i, ch := range num {
        if(i == 0) {
            fmt = append(fmt, '(')
        }
        fmt = append(fmt, ch)
        if (i == 2) {
            fmt = append(fmt, ')', ' ')
        }
        if (i == 5) {
            fmt = append(fmt, '-')
        }
    }
    return string(fmt), nil
}
